### ISSUE TITLE
RFC: omit line number from message

### ISSUE THREAD

**Issue by Seasawher** [leanprover/lean4#4519]
### Proposal

The following output message contains a line number in editor `26:24-39  ?  ?`. But this is quite fragile to change, so I want to remove the line numbering part.

```lean
/--
error: fail to show termination for
  alternate
with errors
argument #2 was not used for structural recursion
  failed to eliminate recursive application
    alternate ys xs

argument #3 was not used for structural recursion
  failed to eliminate recursive application
    alternate ys xs

structural recursion cannot be used

Could not find a decreasing measure.
The arguments relate at each recursive call as follows:
(<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)
            xs ys
1) 26:24-39  ?  ?
Please use `termination_by` to specify a decreasing measure.
-/
#guard_msgs (whitespace := lax) in
def alternate {α : Type} (xs ys : List α) : List α :=
  match xs, ys with
  | [], ys => ys
  | x :: xs, ys => x :: alternate ys xs
```

### Community Feedback

see https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/omit.20line.20number.20from.20message/near/445973308

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others benefit from the changes in this proposal being added, please ask them to add :+1: to it.

**Comment by nomeata** [comment]
I am unsure. If a user is faced with this error message they might really need this help to identify the respective calls, and so far I did not think of a better way. Do you?

I assume it bother you in the context of tests? Happy to accept a PR that adds an option to omit them, similar to `pp.mvars`, that people writing tests with these can turn on.

**Comment by Seasawher** [comment]
> I assume it bother you in the context of tests?

Yes. When writing code to test output messages, the test code must be modified frequently each time the file is changed.

**Comment by nomeata** [comment]
Got it. What is your use case for testing failing function definitions? The lean test suite itself, or some other project?

You could also test it indirectly: use guard_msgs to hide the error text, and then afterwards check that the function is not fully constructed (e.g. no `.eq_def` lemma)

**Comment by Seasawher** [comment]
I'm writing Lean reference book (https://lean-ja.github.io/lean-by-example/, Japanese) and I use `#guard_msgs` to show examples including errors without actual error.

**Comment by Seasawher** [comment]
By the way, since Lean is a highly interactive language and it is possible to see where the error is occurring in the editor, I think it is not necessary to include the location of the error in the message. It would be more convenient for testing if the line number is not displayed by default.

**Comment by Seasawher** [comment]
> You could also test it indirectly: use guard\_msgs to hide the error text, and then afterwards check that the function is not fully constructed (e.g. no `.eq_def` lemma)

oh nice! it is interesting

```lean
#guard_msgs (drop all) in
def alternate {α : Type} (xs ys : List α) : List α :=
  match xs, ys with
  | [], ys => ys
  | x :: xs, ys => x :: alternate ys xs

/-
alternate [1, 2] [3, 4]
-/
#reduce alternate [1, 2] [3, 4]
```

**Comment by nomeata** [comment]
> I'm writing Lean reference book (https://lean-ja.github.io/lean-by-example/, Japanese) and I use `#guard_msgs` to show examples including errors without actual error.

That's great work, thanks! But hopefully you understand that this is a special case and shouldn't guide the error message phrasing.

Are you unblocked by the `drop all` trick? Can we close this then?

**Comment by Seasawher** [comment]
> Are you unblocked by the `drop all` trick?

No.  I would be happy to be able to display the error message itself, so I would still be happy if there was a way to remove the specific line number from the error message.

**Comment by Seasawher** [comment]
> But hopefully you understand that this is a special case and shouldn't guide the error message phrasing.

Ok this is reasonable. You can close this as not planned.
I may be able to achieve what I want to do in the future by using verso or some other method.

**Comment by david-christiansen** [comment]
This is plausible - in Verso, it would be fairly easy to write a program that checked Lean messages for equality modulo substrings that look like source positions and/or metavariable names, so the error in the book matches what users will see but it doesn't need updating quite so often.

**Comment by nomeata** [comment]
No action in two years, probably not happen any time soon.
